File: gestor_parametros.py
from __future__ import annotations

from datetime import date
from decimal import Decimal, InvalidOperation
from typing import Iterable

from modelo_datos import LiquidacionNomina, ParametroNormativo


class ErrorParametro(ValueError):
    """Error de validación o vigencia de un parámetro normativo."""


class GestorParametros:
    PARAMETROS_PORCENTUALES = {
        "PORCENTAJE_ARL_CLASE_I",
        "PORCENTAJE_ARL_CLASE_II",
        "PORCENTAJE_SENA",
        "PORCENTAJE_ICBF",
        "PORCENTAJE_SALUD_TRABAJADOR",
        "PORCENTAJE_SALUD_EMPLEADOR",
        "PORCENTAJE_PENSION_TRABAJADOR",
        "PORCENTAJE_PENSION_EMPLEADOR",
        "PORCENTAJE_FONDO_SOLIDARIDAD",
        "PORCENTAJE_RIESGOS_LABORALES",
        "PORCENTAJE_CAJA_COMPENSACION",
        "PORCENTAJE_BONIFICACION_SERVICIOS_HASTA_TOPE",
        "PORCENTAJE_BONIFICACION_SERVICIOS_SOBRE_TOPE",
        "PORCENTAJE_RETENCION_FUENTE",
    }
    PARAMETROS_MONETARIOS = {
        "SALARIO_MINIMO",
        "VALOR_PUNTO_SALARIAL",
        "VALOR_AUXILIO_TRANSPORTE_VIGENTE",
        "VALOR_HORA_CATEDRA",
        "TOPE_BONIFICACION_SERVICIOS",
        "BASE_MINIMA_RETENCION_FUENTE",
    }

    def __init__(
        self,
        parametros: list[ParametroNormativo],
        liquidaciones: Iterable[LiquidacionNomina] | None = None,
    ) -> None:
        self.parametros = parametros
        self.liquidaciones = list(liquidaciones or [])

    def crear_parametro(self, parametro: ParametroNormativo) -> ParametroNormativo:
        self._validar(parametro)
        if parametro.idParametro is None:
            parametro.idParametro = self._siguiente_id()
        if any(item.idParametro == parametro.idParametro for item in self.parametros):
            raise ErrorParametro(f"Ya existe un parámetro con ID {parametro.idParametro}")
        self._validar_no_solapamiento(parametro)
        parametro.estado = "ACTIVO"
        self.parametros.append(parametro)
        return parametro

    def buscar_parametro_vigente(self, codigo: str, fecha: date | None = None) -> ParametroNormativo | None:
        fecha_consulta = fecha or date.today()
        codigo = self._codigo(codigo)
        candidatos = [
            parametro for parametro in self.parametros
            if self._codigo(parametro.codigo) == codigo
            and self._activo(parametro.estado)
            and self._vigente_en(parametro, fecha_consulta)
        ]
        return max(candidatos, key=lambda parametro: parametro.fechaInicioVigencia or date.min, default=None)

    def modificar_parametro(self, id_parametro: int, **cambios) -> ParametroNormativo:
        actual = self._buscar(id_parametro)
        if self._esta_usado_en_liquidacion(actual):
            raise ErrorParametro("No se puede modificar un parámetro utilizado en una liquidación")
        actualizado = ParametroNormativo(**{campo: cambios.get(campo, getattr(actual, campo)) for campo in actual.__dataclass_fields__})
        actualizado.idParametro = actual.idParametro
        self._validar(actualizado)
        self._validar_no_solapamiento(actualizado, excluir=actual)
        for campo in actual.__dataclass_fields__:
            setattr(actual, campo, getattr(actualizado, campo))
        return actual

    def desactivar_parametro(self, id_parametro: int) -> ParametroNormativo:
        parametro = self._buscar(id_parametro)
        if self._esta_usado_en_liquidacion(parametro):
            raise ErrorParametro("No se puede desactivar un parámetro utilizado en una liquidación")
        parametro.estado = "INACTIVO"
        return parametro

    def listar_parametros_por_tipo(self, tipo_dato: str) -> list[ParametroNormativo]:
        return [item for item in self.parametros if item.tipoDato == tipo_dato]

    def _validar(self, parametro: ParametroNormativo) -> None:
        if not parametro.codigo:
            raise ErrorParametro("El código del parámetro es obligatorio")
        inicio = parametro.fechaInicioVigencia
        fin = parametro.fechaFinVigencia
        if inicio is None:
            raise ErrorParametro("La fecha de inicio de vigencia es obligatoria")
        if fin is not None and fin <= inicio:
            raise ErrorParametro("La fecha de fin debe ser posterior a la fecha de inicio")
        if parametro.valor is None or parametro.valor == "":
            raise ErrorParametro("El valor del parámetro es obligatorio")
        codigo = self._codigo(parametro.codigo)
        try:
            valor = Decimal(parametro.valor)
        except (InvalidOperation, ValueError):
            raise ErrorParametro("El valor del parámetro debe ser numérico") from None
        if codigo in self.PARAMETROS_PORCENTUALES and not Decimal("0") <= valor <= Decimal("1"):
            raise ErrorParametro("Los porcentajes deben estar entre 0 y 1")
        if codigo in self.PARAMETROS_MONETARIOS and valor <= 0:
            raise ErrorParametro("Los montos monetarios deben ser mayores que cero")

    def _validar_no_solapamiento(self, nuevo: ParametroNormativo, excluir: ParametroNormativo | None = None) -> None:
        for actual in self.parametros:
            if actual is excluir or not self._activo(actual.estado) or self._codigo(actual.codigo) != self._codigo(nuevo.codigo):
                continue
            if self._intervalos_se_cruzan(nuevo, actual):
                raise ErrorParametro(f"La vigencia se solapa para el parámetro {self._codigo(nuevo.codigo)}")

    @staticmethod
    def _intervalos_se_cruzan(primero: ParametroNormativo, segundo: ParametroNormativo) -> bool:
        primero_fin = primero.fechaFinVigencia or date.max
        segundo_fin = segundo.fechaFinVigencia or date.max
        return (primero.fechaInicioVigencia or date.min) <= segundo_fin and (segundo.fechaInicioVigencia or date.min) <= primero_fin

    def _esta_usado_en_liquidacion(self, parametro: ParametroNormativo) -> bool:
        codigo = self._codigo(parametro.codigo)
        return any(
            liquidacion.aprobada
            and any(codigo in str(valor or "") for valor in (liquidacion.regimenLiquidado, liquidacion.medioPago, liquidacion.referenciaPago))
            for liquidacion in self.liquidaciones
        )

    def _buscar(self, id_parametro: int) -> ParametroNormativo:
        parametro = next((item for item in self.parametros if item.idParametro == id_parametro), None)
        if parametro is None:
            raise ErrorParametro(f"No existe el parámetro con ID {id_parametro}")
        return parametro

    def _siguiente_id(self) -> int:
        return max((item.idParametro or 0 for item in self.parametros), default=0) + 1

    @staticmethod
    def _codigo(codigo) -> str:
        return str(getattr(codigo, "value", codigo or "")).upper()

    @staticmethod
    def _activo(estado) -> bool:
        return str(getattr(estado, "value", estado)).upper() == "ACTIVO"

    @staticmethod
    def _vigente_en(parametro: ParametroNormativo, fecha: date) -> bool:
        return (parametro.fechaInicioVigencia or date.min) <= fecha and (
            parametro.fechaFinVigencia is None or fecha <= parametro.fechaFinVigencia
        )

File: cpp/gestores/gestor_parametros.cpp
#include "gestor_parametros.h"
#include <ctime>
#include <iomanip>
#include <sstream>
#include <algorithm>

namespace pita {

static const std::unordered_set<std::string> PARAMETROS_PORCENTUALES = {
    "PORCENTAJE_ARL_CLASE_I",
    "PORCENTAJE_ARL_CLASE_II",
    "PORCENTAJE_SENA",
    "PORCENTAJE_ICBF",
    "PORCENTAJE_SALUD_TRABAJADOR",
    "PORCENTAJE_SALUD_EMPLEADOR",
    "PORCENTAJE_PENSION_TRABAJADOR",
    "PORCENTAJE_PENSION_EMPLEADOR",
    "PORCENTAJE_FONDO_SOLIDARIDAD",
    "PORCENTAJE_RIESGOS_LABORALES",
    "PORCENTAJE_CAJA_COMPENSACION",
    "PORCENTAJE_BONIFICACION_SERVICIOS_HASTA_TOPE",
    "PORCENTAJE_BONIFICACION_SERVICIOS_SOBRE_TOPE",
    "PORCENTAJE_RETENCION_FUENTE"
};

static const std::unordered_set<std::string> PARAMETROS_MONETARIOS = {
    "SALARIO_MINIMO",
    "VALOR_PUNTO_SALARIAL",
    "VALOR_AUXILIO_TRANSPORTE_VIGENTE",
    "VALOR_HORA_CATEDRA",
    "TOPE_BONIFICACION_SERVICIOS",
    "BASE_MINIMA_RETENCION_FUENTE"
};

static std::string fecha_hoy() {
    std::time_t t = std::time(nullptr);
    std::tm tm = *std::localtime(&t);
    std::ostringstream oss;
    oss << std::put_time(&tm, "%Y-%m-%d");
    return oss.str();
}

static std::string a_mayusculas(std::string s) {
    std::transform(s.begin(), s.end(), s.begin(), [](unsigned char c) { return std::toupper(c); });
    return s;
}

GestorParametros::GestorParametros(
    ListaEnlazada<ParametroNormativo>& parametros,
    ListaEnlazada<LiquidacionNomina>& liquidaciones
) : parametros(parametros),
    liquidaciones(liquidaciones) {}

int GestorParametros::siguienteId() {
    int maxId = 0;
    for (const auto& p : parametros) {
        if (p.idParametro.has_value() && *p.idParametro > maxId) {
            maxId = *p.idParametro;
        }
    }
    return maxId + 1;
}

ParametroNormativo* GestorParametros::buscar(int idParametro) {
    for (auto& p : parametros) {
        if (p.idParametro.has_value() && *p.idParametro == idParametro) {
            return &p;
        }
    }
    throw ErrorParametro("No existe el parametro con ID " + std::to_string(idParametro));
}

std::string GestorParametros::codigoStr(const ParametroNormativo& p) {
    if (p.codigo.has_value()) {
        return to_string(*p.codigo);
    }
    return "";
}

std::string GestorParametros::codigoStr(const std::string& cod) {
    return a_mayusculas(cod);
}

bool GestorParametros::esActivo(const std::optional<std::string>& estado) {
    if (!estado.has_value()) return false;
    return a_mayusculas(*estado) == "ACTIVO";
}

bool GestorParametros::vigenteEn(const ParametroNormativo& p, const std::string& fecha) {
    std::string inicio = p.fechaInicioVigencia.value_or("0000-00-00");
    if (inicio > fecha) return false;
    if (p.fechaFinVigencia.has_value() && !p.fechaFinVigencia->empty()) {
        if (fecha > *p.fechaFinVigencia) return false;
    }
    return true;
}

bool GestorParametros::intervalosSeCruzan(const ParametroNormativo& primero, const ParametroNormativo& segundo) {
    std::string p_ini = primero.fechaInicioVigencia.value_or("0000-00-00");
    std::string p_fin = (primero.fechaFinVigencia.has_value() && !primero.fechaFinVigencia->empty()) ? *primero.fechaFinVigencia : "9999-99-99";

    std::string s_ini = segundo.fechaInicioVigencia.value_or("0000-00-00");
    std::string s_fin = (segundo.fechaFinVigencia.has_value() && !segundo.fechaFinVigencia->empty()) ? *segundo.fechaFinVigencia : "9999-99-99";

    return (p_ini <= s_fin && s_ini <= p_fin);
}

bool GestorParametros::estaUsadoEnLiquidacion(const ParametroNormativo& p) {
    std::string cod = codigoStr(p);
    for (const auto& l : liquidaciones) {
        if (l.aprobada.value_or(false)) {
            if (l.regimenLiquidado.has_value() && l.regimenLiquidado->find(cod) != std::string::npos) return true;
            if (l.medioPago.has_value() && l.medioPago->find(cod) != std::string::npos) return true;
            if (l.referenciaPago.has_value() && l.referenciaPago->find(cod) != std::string::npos) return true;
        }
    }
    return false;
}

void GestorParametros::validar(const ParametroNormativo& p) {
    if (!p.codigo.has_value()) {
        throw ErrorParametro("El codigo del parametro es obligatorio");
    }
    if (!p.fechaInicioVigencia.has_value() || p.fechaInicioVigencia->empty()) {
        throw ErrorParametro("La fecha de inicio de vigencia es obligatoria");
    }
    if (p.fechaFinVigencia.has_value() && !p.fechaFinVigencia->empty()) {
        if (*p.fechaFinVigencia <= *p.fechaInicioVigencia) {
            throw ErrorParametro("La fecha de fin debe ser posterior a la fecha de inicio");
        }
    }
    if (!p.valor.has_value() || p.valor->empty()) {
        throw ErrorParametro("El valor del parametro es obligatorio");
    }

    double v = 0.0;
    try {
        v = std::stod(*p.valor);
    } catch (...) {
        throw ErrorParametro("El valor del parametro debe ser numerico");
    }

    std::string cod = codigoStr(p);
    if (PARAMETROS_PORCENTUALES.find(cod) != PARAMETROS_PORCENTUALES.end()) {
        if (v < 0.0 || v > 1.0) {
            throw ErrorParametro("Los porcentajes deben estar entre 0 y 1");
        }
    }
    if (PARAMETROS_MONETARIOS.find(cod) != PARAMETROS_MONETARIOS.end()) {
        if (v <= 0.0) {
            throw ErrorParametro("Los montos monetarios deben ser mayores que cero");
        }
    }
}

void GestorParametros::validarNoSolapamiento(const ParametroNormativo& nuevo, const ParametroNormativo* excluir) {
    std::string nuevoCod = codigoStr(nuevo);
    for (const auto& actual : parametros) {
        if (&actual == excluir) continue;
        if (!esActivo(actual.estado)) continue;
        if (codigoStr(actual) != nuevoCod) continue;

        if (intervalosSeCruzan(nuevo, actual)) {
            throw ErrorParametro("La vigencia se solapa para el parametro " + nuevoCod);
        }
    }
}

ParametroNormativo& GestorParametros::crearParametro(ParametroNormativo parametro) {
    validar(parametro);
    if (!parametro.idParametro.has_value()) {
        parametro.idParametro = siguienteId();
    } else {
        for (const auto& item : parametros) {
            if (item.idParametro.has_value() && *item.idParametro == *parametro.idParametro) {
                throw ErrorParametro("Ya existe un parametro con ID " + std::to_string(*parametro.idParametro));
            }
        }
    }
    validarNoSolapamiento(parametro);
    parametro.estado = "ACTIVO";
    parametros.push_back(std::move(parametro));
    return parametros.back();
}

ParametroNormativo* GestorParametros::buscarParametroVigente(const std::string& codigo, const std::string& fecha) {
    std::string fechaConsulta = fecha.empty() ? fecha_hoy() : fecha;
    std::string cod = codigoStr(codigo);

    ParametroNormativo* mejor = nullptr;
    for (auto& p : parametros) {
        if (codigoStr(p) == cod && esActivo(p.estado) && vigenteEn(p, fechaConsulta)) {
            if (!mejor || (p.fechaInicioVigencia.value_or("") > mejor->fechaInicioVigencia.value_or(""))) {
                mejor = &p;
            }
        }
    }
    return mejor;
}

ParametroNormativo* GestorParametros::obtenerParametro(const std::string& codigo, const std::string& fecha) {
    return buscarParametroVigente(codigo, fecha);
}

std::optional<double> GestorParametros::obtenerParametroVigente(const std::string& codigo, const std::string& fecha) {
    ParametroNormativo* p = buscarParametroVigente(codigo, fecha);
    if (p && p->valor.has_value() && !p->valor->empty()) {
        try {
            return std::stod(*p->valor);
        } catch (...) {
            return std::nullopt;
        }
    }
    return std::nullopt;
}

ParametroNormativo& GestorParametros::modificarParametro(int idParametro, const ParametroNormativo& datosActualizados) {
    ParametroNormativo* actual = buscar(idParametro);
    if (estaUsadoEnLiquidacion(*actual)) {
        throw ErrorParametro("No se puede modificar un parametro utilizado en una liquidacion");
    }

    ParametroNormativo copia = datosActualizados;
    copia.idParametro = actual->idParametro;
    validar(copia);
    validarNoSolapamiento(copia, actual);

    *actual = copia;
    return *actual;
}

ParametroNormativo& GestorParametros::desactivarParametro(int idParametro) {
    ParametroNormativo* p = buscar(idParametro);
    if (estaUsadoEnLiquidacion(*p)) {
        throw ErrorParametro("No se puede desactivar un parametro utilizado en una liquidacion");
    }
    p->estado = "INACTIVO";
    return *p;
}

ListaEnlazada<ParametroNormativo> GestorParametros::listarParametrosPorTipo(const std::string& tipoDato) const {
    ListaEnlazada<ParametroNormativo> res;
    for (const auto& p : parametros) {
        if (p.tipoDato.has_value() && *p.tipoDato == tipoDato) {
            res.push_back(p);
        }
    }
    return res;
}

} // namespace pita
